<div>
<img src=https://www.institutedata.com/wp-content/uploads/2019/10/iod_h_tp_primary_c.svg width="300">
</div>

# Lab 2.1.1
# *Data Wrangling and Munging with Pandas*

## Part 1: Wrangling Data

The term "data wrangling" is analogous to capturing wild horses and getting them into a fenced area; the horses are data and the fencing is your computer. The more common data wrangling tasks include:

- reading flat files
- reading Excel files
- downloading from web pages
  - csv
  - html
  - json

In [2]:
import numpy as np
import pandas as pd

*It is good practice to display the library version numbers for future reference:*

In [3]:
print('Numpy: ', np.__version__)
print('Pandas: ', pd.__version__)

Numpy:  1.24.3
Pandas:  2.0.3


### CSV Files

Below are three attempts to load the file "bikeshare.csv" into a DataFrame named `bikes`. Why are they wrong?

In [3]:
# wrong:
bikes = pd.read_table(r"C:\Users\OmarZazou\OneDrive - Ediom Pty Ltd\Documents\Resources\Data Science Course\Course Material\DATA\bikeshare.csv", header = None)
print(bikes.head())

# wrong:
bikes = pd.read_table(r"C:\Users\OmarZazou\OneDrive - Ediom Pty Ltd\Documents\Resources\Data Science Course\Course Material\DATA\bikeshare.csv", header = 1)
print(bikes.head())

# wrong:
bikes = pd.read_table(r"C:\Users\OmarZazou\OneDrive - Ediom Pty Ltd\Documents\Resources\Data Science Course\Course Material\DATA\bikeshare.csv", header = 0)
print(bikes.head())

                                                   0
0  instant,dteday,season,yr,mnth,hr,holiday,weekd...
1  1,2011-01-01,1,0,1,0,0,6,0,1,0.24,0.2879,0.81,...
2  2,2011-01-01,1,0,1,1,0,6,0,1,0.22,0.2727,0.8,0...
3  3,2011-01-01,1,0,1,2,0,6,0,1,0.22,0.2727,0.8,0...
4  4,2011-01-01,1,0,1,3,0,6,0,1,0.24,0.2879,0.75,...
  1,2011-01-01,1,0,1,0,0,6,0,1,0.24,0.2879,0.81,0,3,13,16
0  2,2011-01-01,1,0,1,1,0,6,0,1,0.22,0.2727,0.8,0...     
1  3,2011-01-01,1,0,1,2,0,6,0,1,0.22,0.2727,0.8,0...     
2  4,2011-01-01,1,0,1,3,0,6,0,1,0.24,0.2879,0.75,...     
3  5,2011-01-01,1,0,1,4,0,6,0,1,0.24,0.2879,0.75,...     
4  6,2011-01-01,1,0,1,5,0,6,0,2,0.24,0.2576,0.75,...     
  instant,dteday,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0  1,2011-01-01,1,0,1,0,0,6,0,1,0.24,0.2879,0.81,...                                                                   
1  2,2011-01-01,1,0,1,1,0,6,0,1,0.22,0.2727,0.8,0...                                          

All 3 instances have the mistake of not defining the delimiter therefore the table is structured incorrectly because it defaults to no delimiter.

The first instance has an additional mistake where we are telling the method that there is no header so the first row is interpreted as an item/object.
The second instance has an additional mistake where we are defining the header as the second row in the table (since python indices start at 0, not 1).
The third instance has no additional mistake, it would have worked had we defined the correct delimiter.

ANSWER:

Load the file "bikeshare.csv" into a DataFrame named `bikes`, and confirm that it was loaded properly:

In [4]:
#ANSWER:
bikes = pd.read_table(r"C:\Users\OmarZazou\OneDrive - Ediom Pty Ltd\Documents\Resources\Data Science Course\Course Material\DATA\bikeshare.csv", delimiter=',')
print(bikes.head())

   instant      dteday  season  yr  mnth  hr  holiday  weekday  workingday  \
0        1  2011-01-01       1   0     1   0        0        6           0   
1        2  2011-01-01       1   0     1   1        0        6           0   
2        3  2011-01-01       1   0     1   2        0        6           0   
3        4  2011-01-01       1   0     1   3        0        6           0   
4        5  2011-01-01       1   0     1   4        0        6           0   

   weathersit  temp   atemp   hum  windspeed  casual  registered  cnt  
0           1  0.24  0.2879  0.81        0.0       3          13   16  
1           1  0.22  0.2727  0.80        0.0       8          32   40  
2           1  0.22  0.2727  0.80        0.0       5          27   32  
3           1  0.24  0.2879  0.75        0.0       3          10   13  
4           1  0.24  0.2879  0.75        0.0       0           1    1  


Note that we could have used `read.csv()` above. When is `read_table()` necessary?

?:
ANSWER: When `sep` is not the comma character, or we need fine control that `read.csv()` does not provide.

Flat files can be full of surprises. Here are some issues to watch out for:

- separator character is something other than the comma
  - ";", "|", and tab are popular
- newline character is something other than what the O/S expects
  - Tip: Don't hard-code the character codes for carriage returns, linefeeds, etc. Use Python's built-in representation instead (e.g. Python translates "\n" to the newline character and "\t" to the tab character on any O/S).
- truncated lines
  - if there are empty fields at the end of a line it is possible that their separators will be missing, resulting in a "jagged" file
- embedded commas or quotes
  - a free-text field containing embedded commas may split into separate fields on input
  - a free-text field containing embedded quotes may not parse correctly
- unescaped characters
  - the "\" character indicates a control code to Python, which will break the I/O
    - e.g. the substring "\u0123" will be interpreted as Unicode(0123) -- which may not be what the file creator intended
  - these may need to be fixed by loading whole strings and then parsing into a new data frame
  
Tip: Most issues can be dealt with by correctly specifying the parameters of the function you use to load the file. Read the doco before reading the data!

### Reading Excel Files

In [5]:
from pandas import ExcelFile  # Nb. Need to install xlrd from conda (it does not automatically install with pandas)

In [6]:
df = pd.read_excel(r'C:\Users\OmarZazou\OneDrive - Ediom Pty Ltd\Documents\Resources\Data Science Course\Course Material\DATA\Iris.xls', sheet_name = 'Data' )
df

ImportError: Missing optional dependency 'xlrd'. Install xlrd >= 2.0.1 for xls Excel support Use pip or conda to install xlrd.

It is usually better to load data correctly than to meddle with the source file or load it 'warts and all' and then try to parse it in code. The Pandas functions for reading files have parameters that provide the control we need. For example, we could make multiple calls to `read_excel()`, using combinations of the `header`, `usecols`, `skiprows`, `nrows`, and `skipfooter` parameters to load one table at a time from a spreadsheet with multiple tables.

Load the above file without the unwanted columns:

In [7]:
#ANSWER
df = pd.read_excel(r'C:\Users\OmarZazou\OneDrive - Ediom Pty Ltd\Documents\Resources\Data Science Course\Course Material\DATA\Iris.xls', sheet_name = 'Data', usecols = [1, 2, 4] )
df

Petal_width  Petal_length  Sepal_length
0            0.2           1.4           5.1
1            0.2           1.4           4.9
2            0.2           1.3           4.7
3            0.2           1.5           4.6
4            0.2           1.4           5.0
..           ...           ...           ...
145          2.3           5.2           6.7
146          1.9           5.0           6.3
147          2.0           5.2           6.5
148          2.3           5.4           6.2
149          1.8           5.1           5.9

[150 rows x 3 columns]

### Importing Data Directly from the Web

We usually want to store a local copy of a data file that we download from the Web, but when data retention is not a priority it is convenient to download the data directly into our running Python environment.

#### Importing Text Files from the Web

The web is the 'wild west' of data formats. However, we can usually expect good behaviour from files that are automatically generated by a service, such as the earthquake report:

In [8]:
df = pd.read_csv('https://earthquake.usgs.gov/earthquakes/feed/v1.0/summary/2.5_hour.csv')
df.head()

time  latitude  longitude    depth  mag magType  nst  \
0  2023-12-09T03:59:19.627Z  -23.0534   -68.2748  137.257  4.8      mb   44   
1  2023-12-09T03:59:15.181Z   -6.0529   150.1657   51.402  4.8      mb   25   

   gap   dmin   rms  ...                   updated  \
0   98  0.134  0.82  ...  2023-12-09T04:14:46.040Z   
1  129  2.717  0.56  ...  2023-12-09T04:18:27.040Z   

                                      place        type horizontalError  \
0  17 km SSW of San Pedro de Atacama, Chile  earthquake            7.33   
1        55 km S of Kimbe, Papua New Guinea  earthquake            5.66   

  depthError  magError  magNst    status  locationSource magSource  
0      4.395     0.089      39  reviewed              us        us  
1      9.613     0.135      17  reviewed              us        us  

[2 rows x 22 columns]

#### Importing HTML Files from the Web

Working with unstructured HTML files relies heavily on library functions. This one, however, is well-structured:

In [9]:
url = 'https://www.ccra.com/airport-codes/'

df = pd.read_html(url)
df

[                City      Country Code
 0            Aalborg      Denmark  AAL
 1           Aalesund       Norway  AES
 2             Aarhus      Denmark  AAR
 3     Abbotsford, BC       Canada  YXX
 4     Abbotsford, BC       Canada  YXX
 ...              ...          ...  ...
 1574        Zanzibar     Tanzania  ZNZ
 1575        Zaragoza        Spain  ZAZ
 1576       Zhengzhou        China  CGO
 1577        Zhoushan        China  HSN
 1578          Zurich  Switzerland  ZRH
 
 [1579 rows x 3 columns]]

In [10]:
df[0]

City      Country Code
0            Aalborg      Denmark  AAL
1           Aalesund       Norway  AES
2             Aarhus      Denmark  AAR
3     Abbotsford, BC       Canada  YXX
4     Abbotsford, BC       Canada  YXX
...              ...          ...  ...
1574        Zanzibar     Tanzania  ZNZ
1575        Zaragoza        Spain  ZAZ
1576       Zhengzhou        China  CGO
1577        Zhoushan        China  HSN
1578          Zurich  Switzerland  ZRH

[1579 rows x 3 columns]

#### Importing XML Files from the Web

XML files are semi-structured, but you're at the mercy of the file creator. If every record has the same format it will be much easier, but practical applications often require a lot of custom code. Here are a few examples: https://pandas.pydata.org/docs/user_guide/io.html#io-read-xml

#### Importing JSON Files from the Web

Like XML, JSON files are semi-structured and may require work to capture the schema into a dataframe. Here is a simple example:

In [7]:
url = 'https://microsoftedge.github.io/Demos/json-dummy-data/64KB.json'

# Load the first sheet of the JSON file into a data frame
df = pd.read_json(url, orient = 'columns')
df.head()

name language                id  \
0  Adeel Solangi   Sindhi  V59OF92YF627HFY0   
1  Afzal Ghaffar   Sindhi  ENTOCR13RSCLZ6KU   
2  Aamir Solangi   Sindhi  IAKPO3R4761JDRVG   
3  Abla Dilmurat   Uyghur  5ZVOEPMJUI4MB4EN   
4       Adil Eli   Uyghur  6VTI8X6LL0MMPJCC   

                                                 bio  version  
0  Donec lobortis eleifend condimentum. Cras dict...     6.10  
1  Aliquam sollicitudin ante ligula, eget malesua...     1.88  
2  Vestibulum pharetra libero et velit gravida eu...     7.27  
3  Donec lobortis eleifend condimentum. Morbi ac ...     2.53  
4  Vivamus id faucibus velit, id posuere leo. Mor...     6.49

## Part 2: Data Munging

Data munging is manipulating data to get it into a form that we can start running analyses on (which usually means getting the data into a DataFrame). Before we get to this stage, we may need to remove headers or footers, transpose columns to rows, split wide data tables into long ones, and so on. (Nb. Excel files can be particularly troublesome, because users can format their data in mixed, complex shapes.) Essentially, we need to follow Hadley Wickham's guidelines for tidy datasets (http://vita.had.co.nz/papers/tidy-data.html):

The end goal of the cleaning data process:

- each variable should be in one column
- each observation should comprise one row
- each type of observational unit should form one table
- include key columns for linking multiple tables
- the top row contains (sensible) variable names
- in general, save data as one file per table


### Dataset Morphology

Once we have our dataset in a DataFrame (or Series, if our data is only 1-dimensional), we can start examining its size and content.

How many rows and columns are in `bikes`?

In [8]:
#ANSWER
bikes.shape

(17379, 17)

What are the column names in `bikes`?

In [9]:
#ANSWER
bikes.columns

Index(['instant', 'dteday', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday',
       'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed',
       'casual', 'registered', 'cnt'],
      dtype='object')

What are the data types of these columns?

In [14]:
#ANSWER
bikes.dtypes

instant         int64
dteday         object
season          int64
yr              int64
mnth            int64
hr              int64
holiday         int64
weekday         int64
workingday      int64
weathersit      int64
temp          float64
atemp         float64
hum           float64
windspeed     float64
casual          int64
registered      int64
cnt             int64
dtype: object

What is the (row) index for this DataFrame?

In [15]:
#ANSWER
bikes.index

RangeIndex(start=0, stop=17379, step=1)

The index is a range starting at 0 and ending at 17379 with a step of 1.

https://www.dataquest.io/blog/python-json-tutorial/

## Slicing and Dicing

It is often preferable to refer to DataFrame columns by name, but there is more than one way to do this.
Do `bikes['season']` and `bikes[['season']]` give the same object? Demonstrate:

In [16]:
#ANSWER
bikes['season']

0        1
1        1
2        1
3        1
4        1
        ..
17374    1
17375    1
17376    1
17377    1
17378    1
Name: season, Length: 17379, dtype: int64

In [17]:
bikes[['season']]

season
0           1
1           1
2           1
3           1
4           1
...       ...
17374       1
17375       1
17376       1
17377       1
17378       1

[17379 rows x 1 columns]

The single bracket notation returns a series/set of values while the double bracket notation returns a datfarame object.

How would we use object notation to show the first 4 rows of `atemp`?

In [40]:
#ANSWER
bikes.atemp[:4]

0    0.2879
1    0.2727
2    0.2727
3    0.2879
Name: atemp, dtype: float64

Algorithms that loop over multiple columns often access DataFrame columns by index. However, none of the following work (try them out by uncommenting / removing the "#E: " ):

In [21]:
#bikes[[0]]
#E: bikes[0]
#E: bikes[0,0]
#E: bikes[[0,0]]

KeyError: "None of [Index([0, 0], dtype='int32')] are in the [columns]"

What is the correct way to access the 1st row of the DataFrame by its index?

In [24]:
#ANSWER
bikes[:1]

instant      dteday  season  yr  mnth  hr  holiday  weekday  workingday  \
0        1  2011-01-01       1   0     1   0        0        6           0   

   weathersit  temp   atemp   hum  windspeed  casual  registered  cnt  
0           1  0.24  0.2879  0.81        0.0       3          13   16

What is the correct way to access the 2nd column of the DataFrame by its index?

In [28]:
#ANSWER
bikes.iloc[1:2]

instant      dteday  season  yr  mnth  hr  holiday  weekday  workingday  \
1        2  2011-01-01       1   0     1   1        0        6           0   

   weathersit  temp   atemp  hum  windspeed  casual  registered  cnt  
1           1  0.22  0.2727  0.8        0.0       8          32   40

## Handling Missing Values

What is the Pandas `isnull` function for?

ANSWER: The Pandas isnull function is a function which loops through the values it'a spplied to (may be a whole dataframe or a column from the dataframe) and returns a Boolean (True or False) for each calue depending on whether the value is NaN/empty/null or not. It returns False for a value which is not empty and True for a value which is empty.

We can apply `isnull` to the `bikes` DataFrame to show the result for every element:

In [29]:
bikes.isnull().head()

instant  dteday  season     yr   mnth     hr  holiday  weekday  workingday  \
0    False   False   False  False  False  False    False    False       False   
1    False   False   False  False  False  False    False    False       False   
2    False   False   False  False  False  False    False    False       False   
3    False   False   False  False  False  False    False    False       False   
4    False   False   False  False  False  False    False    False       False   

   weathersit   temp  atemp    hum  windspeed  casual  registered    cnt  
0       False  False  False  False      False   False       False  False  
1       False  False  False  False      False   False       False  False  
2       False  False  False  False      False   False       False  False  
3       False  False  False  False      False   False       False  False  
4       False  False  False  False      False   False       False  False

However, we usually start at a higher level. How many nulls are in `bikes` altogether?

In [35]:
#ANSWER
bikes.isnull().sum()

instant       0
dteday        0
season        0
yr            0
mnth          0
hr            0
holiday       0
weekday       0
workingday    0
weathersit    0
temp          0
atemp         0
hum           0
windspeed     0
casual        0
registered    0
cnt           0
dtype: int64

If this result were nonzero we would next want to find out which columns contained nulls. How can this be done in one line of code?

ANSWER: We can already see which columns have nulls from the line of code above.
If we wante dto find out which specific values within a column were null we can use the line of code: bikes[bikes.column_name.isnull()].

In [46]:
#ANSWER
bikes.workingday.unique()

array([0, 1], dtype=int64)

What is the Numpy object `nan` used for? (Write a descriptive answer.)

ANSWER: It stands for Not a Number and it is used to imply there is no numerical value contained in this object.

Write (and verify) a function that performs scalar division with built-in handling of the edge case (i.e. return a value instead of just trapping the error):

In [11]:
#ANSWER
def return_nan(num, den):
    try:
        x=num/den
    except:
        x = np.nan
    return x

In [15]:
return_nan('korm',5)

nan

Apply the Pandas `isna` function to the following data objects:

In [37]:
x = 2.3
y = np.nan
print(x, y)

2.3 nan


In [40]:
#ANSWER
pd.isna(x)

False

In [41]:
pd.isna(y)

True

In [43]:
array = np.array([[1, np.nan, 3], [4, 5, np.nan]])
print(array)

[[ 1. nan  3.]
 [ 4.  5. nan]]


In [46]:
#ANSWER
pd.isna(array)

array([[False,  True, False],
       [False, False,  True]])

How is the pandas I/O parameter `na_values` used?

? ANSWER:

## Data Profiling

### Counts

When there are categorical variables in a dataset we will want to know how many possible values there are in each column. (Nb. If the dataset is a sample of a larger one, our sample may not capture all possible values of every categorical.)

How many (different) seasons are in `bikes`?

In [16]:
#ANSWER
bikes.season.unique()

array([1, 2, 3, 4], dtype=int64)

We have 4 different seasons in bikes.

In [51]:
bikes.season.value_counts()

season
3    4496
2    4409
1    4242
4    4232
Name: count, dtype: int64

In [52]:
bikes.season.value_counts(normalize=True)

season
3    0.258703
2    0.253697
1    0.244088
4    0.243512
Name: proportion, dtype: float64

### Ranges

Print the range of the `instant`, `dteday`, and `windspeed` columns:

In [37]:
#ANSWER
print('The range for instant is from ' + str(bikes.instant.min()) + ' to ' + str(bikes.instant.max()))

The range for instant is from 1 to 17379


In [38]:
print('The range for dteday is from ' + str(bikes.dteday.min()) + ' to ' + str(bikes.dteday.max()))

The range for dteday is from 2011-01-01 to 2012-12-31


In [39]:
print('The range for instant is from ' + str(bikes.windspeed.min()) + ' to ' + str(bikes.windspeed.max()))

The range for instant is from 0.0 to 0.8507


Compute and print the overall minimum and maximum of the numeric data columns:

In [55]:
bikes_min, bikes_max = (min(bikes.min(numeric_only=True)), max(bikes.max(numeric_only=True)))
bikes_min, bikes_max

(0.0, 17379.0)

### Quantiles

Pandas makes computing quantiles easy. This is how to get the median of a Series:

In [40]:
bikes['atemp'].quantile(0.5)

0.4848

Of course, the `quantiles` method can take a tuple as its argument. Compute the 10th, 25th, 50th, 75th, and 90th percentiles in one line of code:

In [41]:
#ANSWER
bikes['atemp'].quantile((0.1, 0.25, 0.5, 0.75,0.9))

0.10    0.2424
0.25    0.3333
0.50    0.4848
0.75    0.6212
0.90    0.6970
Name: atemp, dtype: float64

### Cuts

Sometimes we want to split the sample not by the quantiles of the distribution but by the range of the data. Let's take a closer look at `atemp`:

In [42]:
type(bikes['atemp'])

pandas.core.series.Series

In [43]:
bikes.sample(5)

instant      dteday  season  yr  mnth  hr  holiday  weekday  \
13374    13375  2012-07-16       3   1     7  11        0        1   
3245      3246  2011-05-20       2   0     5   3        0        5   
5910      5911  2011-09-08       3   0     9  19        0        4   
15230    15231  2012-10-01       4   1    10  19        0        1   
14930    14931  2012-09-19       3   1     9   7        0        3   

       workingday  weathersit  temp   atemp   hum  windspeed  casual  \
13374           1           1  0.84  0.8030  0.53     0.2239      57   
3245            1           1  0.44  0.4394  1.00     0.0000       1   
5910            1           3  0.62  0.5152  1.00     0.3284       1   
15230           1           2  0.54  0.5152  0.68     0.1045      33   
14930           1           1  0.52  0.5000  0.63     0.2239      16   

       registered  cnt  
13374         125  182  
3245            4    5  
5910           51   52  
15230         483  516  
14930         529  545

Suppose we decide to sort these values into 4 bins of equal width, but we want to apply the resulting groups to the entire DataFrame. Basically, we need to add a row label that indicates which bin each sample belongs in. Let's call this label "atemp_level", and use the `cut` method to populate it:

In [45]:
atemp_level = pd.cut(bikes['atemp'], bins = 4)

What is `atemp_level`?

In [49]:
#ANSWER
type(atemp_level)

pandas.core.series.Series

Here is a random sample of `atemp_level`:

In [50]:
atemp_level.sample(5)

11504    (0.25, 0.5]
1807     (0.25, 0.5]
15508    (0.25, 0.5]
7999     (0.25, 0.5]
7115     (0.25, 0.5]
Name: atemp, dtype: category
Categories (4, interval[float64, right]): [(-0.001, 0.25] < (0.25, 0.5] < (0.5, 0.75] < (0.75, 1.0]]

So, by default, `cut` produces labels that indicate the bin boundaries for each element in the series it was applied to. Usually, we will specify labels that are appropriate to the discretisation we are applying:

In [51]:
atemp_level = pd.cut(bikes['atemp'], bins = 4, labels = ["cool", "mild", "warm", "hot"])
atemp_level.sample(5)

372      cool
12881    warm
4386     warm
3754     warm
11404    mild
Name: atemp, dtype: category
Categories (4, object): ['cool' < 'mild' < 'warm' < 'hot']

Incorporate the new `atemp_level` column into the `bikes` DataFrame and use it to count the number of "mild" `atemp` entries in `season` 2:

In [52]:
#ANSWER
bikes['atemp_level'] = atemp_level
bikes

instant      dteday  season  yr  mnth  hr  holiday  weekday  \
0            1  2011-01-01       1   0     1   0        0        6   
1            2  2011-01-01       1   0     1   1        0        6   
2            3  2011-01-01       1   0     1   2        0        6   
3            4  2011-01-01       1   0     1   3        0        6   
4            5  2011-01-01       1   0     1   4        0        6   
...        ...         ...     ...  ..   ...  ..      ...      ...   
17374    17375  2012-12-31       1   1    12  19        0        1   
17375    17376  2012-12-31       1   1    12  20        0        1   
17376    17377  2012-12-31       1   1    12  21        0        1   
17377    17378  2012-12-31       1   1    12  22        0        1   
17378    17379  2012-12-31       1   1    12  23        0        1   

       workingday  weathersit  temp   atemp   hum  windspeed  casual  \
0               0           1  0.24  0.2879  0.81     0.0000       3   
1               0           1  0.22  0.2727  0.80     0.0000       8   
2               0           1  0.22  0.2727  0.80     0.0000       5   
3               0           1  0.24  0.2879  0.75     0.0000       3   
4               0           1  0.24  0.2879  0.75     0.0000       0   
...           ...         ...   ...     ...   ...        ...     ...   
17374           1           2  0.26  0.2576  0.60     0.1642      11   
17375           1           2  0.26  0.2576  0.60     0.1642       8   
17376           1           1  0.26  0.2576  0.60     0.1642       7   
17377           1           1  0.26  0.2727  0.56     0.1343      13   
17378           1           1  0.26  0.2727  0.65     0.1343      12   

       registered  cnt atemp_level  
0              13   16        mild  
1              32   40        mild  
2              27   32        mild  
3              10   13        mild  
4               1    1        mild  
...           ...  ...         ...  
17374         108  119        mild  
17375          81   89        mild  
17376          83   90        mild  
17377          48   61        mild  
17378          37   49        mild  

[17379 rows x 18 columns]

In [53]:
bikes[(bikes.atemp_level == 'mild') & (bikes.season == 2)].instant.count()

1829

*Nb. The `atemp_level` variable we created is what the R language calls a "factor". Pandas has introduced a new data type called "category" that is similar to R's factors.*

# Synthetic Data

Sometimes we may want to generate test data, or we may need to initialise a series, matrix, or data frame for input to an algorithm. Numpy has several methods we can use for this.

Execute the following, then check the shape and content of each variable:

In [60]:
# Creating arrays with initial values
a = np.zeros((3))
b = np.ones((1,3))
c = np.random.randint(1,10,(2,3,4))   # randint(low, high, size)
d = np.arange(4)
e = np.array( [[1,2,3,4], [5,6,7,8]] )
print(a)
print(b)
print(c)
print(d)
print(e)

[0. 0. 0.]
[[1. 1. 1.]]
[[[2 4 5 4]
  [3 9 3 2]
  [1 7 1 7]]

 [[7 3 6 1]
  [5 5 8 6]
  [2 9 1 4]]]
[0 1 2 3]
[[1 2 3 4]
 [5 6 7 8]]


## Load Data

Load rock.csv and clean the dataset.

In [69]:
# ANSWER
rock_csv = r"C:\Users\OmarZazou\OneDrive - Ediom Pty Ltd\Documents\Resources\Data Science Course\Course Material\DATA\rock.csv"
rock = pd.read_csv(rock_csv)
rock

Song Clean ARTIST CLEAN Release Year  \
0                                Caught Up in You  .38 Special         1982   
1                                    Fantasy Girl  .38 Special          NaN   
2                                 Hold On Loosely  .38 Special         1981   
3                          Rockin' Into the Night  .38 Special         1980   
4                               Art For Arts Sake         10cc         1975   
...                                           ...          ...          ...   
2225                      She Loves My Automobile       ZZ Top          NaN   
2226                            Tube Snake Boogie       ZZ Top         1981   
2227                                         Tush       ZZ Top         1975   
2228                                   TV Dinners       ZZ Top         1983   
2229  WAITIN' FOR THE BUS/JESUS JUST LEFT CHICAGO       ZZ Top         1973   

                                               COMBINED  First?  Year?  \
0                       Caught Up in You by .38 Special       1      1   
1                           Fantasy Girl by .38 Special       1      0   
2                        Hold On Loosely by .38 Special       1      1   
3                 Rockin' Into the Night by .38 Special       1      1   
4                             Art For Arts Sake by 10cc       1      1   
...                                                 ...     ...    ...   
2225                  She Loves My Automobile by ZZ Top       1      0   
2226                        Tube Snake Boogie by ZZ Top       1      1   
2227                                     Tush by ZZ Top       1      1   
2228                               TV Dinners by ZZ Top       1      1   
2229  WAITIN' FOR THE BUS/JESUS JUST LEFT CHICAGO by...       1      1   

      PlayCount  F*G  
0            82   82  
1             3    0  
2            85   85  
3            18   18  
4             1    1  
...         ...  ...  
2225          1    0  
2226         32   32  
2227        109  109  
2228          1    1  
2229          2    2  

[2230 rows x 8 columns]

## Check Column Names

Check column names and clean.

In [70]:
# ANSWER
rock.columns

Index(['Song Clean', 'ARTIST CLEAN', 'Release Year', 'COMBINED', 'First?',
       'Year?', 'PlayCount', 'F*G'],
      dtype='object')

## Replace Null Values With 0

Check 'release_year' column whether this column has any null values or not. Replace null values with 0.

In [71]:
# ANSWER
rock.isnull().sum()

Song Clean        0
ARTIST CLEAN      0
Release Year    577
COMBINED          0
First?            0
Year?             0
PlayCount         0
F*G               0
dtype: int64

In [72]:
rock['Release Year'].fillna(value=0, inplace=True)

In [73]:
rock.isnull().sum()

Song Clean      0
ARTIST CLEAN    0
Release Year    0
COMBINED        0
First?          0
Year?           0
PlayCount       0
F*G             0
dtype: int64

## Check Datatypes of Dataset

Check datatypes of the dataset. Is there any column which should be int instead of object? Fix the column.

In [74]:
# ANSWER
rock.dtypes

Song Clean      object
ARTIST CLEAN    object
Release Year    object
COMBINED        object
First?           int64
Year?            int64
PlayCount        int64
F*G              int64
dtype: object

In [75]:
rock['Release Year'].unique()

array(['1982', 0, '1981', '1980', '1975', '2000', '2002', '1992', '1985',
       '1993', '1976', '1995', '1979', '1984', '1977', '1990', '1986',
       '1974', '2014', '1987', '1973', '2001', '1989', '1997', '1971',
       '1972', '1994', '1970', '1966', '1965', '1983', '1955', '1978',
       '1969', '1999', '1968', '1988', '1962', '2007', '1967', '1958',
       '1071', '1996', '1991', '2005', '2011', '2004', '2012', '2003',
       '1998', '2008', '1964', '2013', '2006', 'SONGFACTS.COM', '1963',
       '1961'], dtype=object)

In [76]:
rock['Release Year'].value_counts()

Release Year
0                577
1973             104
1977              83
1975              83
1970              81
1971              75
1969              72
1980              70
1978              64
1979              63
1981              61
1967              61
1983              60
1976              56
1982              54
1984              51
1972              50
1974              48
1968              46
1985              39
1987              39
1986              37
1991              34
1989              32
1966              30
1988              29
1965              28
1994              25
1990              22
1993              19
1964              14
1992              14
1999              13
1995              10
1997               9
1963               9
1996               9
1998               6
2002               6
2005               5
2004               5
2012               5
2001               4
2007               3
1962               3
2000               3
2011               3


In [77]:
change = rock['Release Year'] == 'SONGFACTS.COM'
rock[change] = 0

In [78]:
rock['Release Year'].unique()

array(['1982', 0, '1981', '1980', '1975', '2000', '2002', '1992', '1985',
       '1993', '1976', '1995', '1979', '1984', '1977', '1990', '1986',
       '1974', '2014', '1987', '1973', '2001', '1989', '1997', '1971',
       '1972', '1994', '1970', '1966', '1965', '1983', '1955', '1978',
       '1969', '1999', '1968', '1988', '1962', '2007', '1967', '1958',
       '1071', '1996', '1991', '2005', '2011', '2004', '2012', '2003',
       '1998', '2008', '1964', '2013', '2006', '1963', '1961'],
      dtype=object)

In [79]:
rock['Release Year'] = rock["Release Year"].apply(pd.to_numeric)

In [80]:
rock.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Song Clean    2230 non-null   object
 1   ARTIST CLEAN  2230 non-null   object
 2   Release Year  2230 non-null   int64 
 3   COMBINED      2230 non-null   object
 4   First?        2230 non-null   int64 
 5   Year?         2230 non-null   int64 
 6   PlayCount     2230 non-null   int64 
 7   F*G           2230 non-null   int64 
dtypes: int64(5), object(3)
memory usage: 139.5+ KB


## Check Min, Max of Each Column

Is there any illogical value in any column? How can we fix that?

In [81]:
# ANSWER
rock.describe(include=[np.number]).T

count         mean         std  min  25%     50%     75%  \
Release Year  2230.0  1465.331390  867.196161  0.0  0.0  1973.0  1981.0   
First?        2230.0     0.999552    0.021176  0.0  1.0     1.0     1.0   
Year?         2230.0     0.740807    0.438290  0.0  0.0     1.0     1.0   
PlayCount     2230.0    16.872197   25.303262  0.0  1.0     4.0    21.0   
F*G           2230.0    15.047982   25.288624  0.0  0.0     3.0    18.0   

                 max  
Release Year  2014.0  
First?           1.0  
Year?            1.0  
PlayCount      142.0  
F*G            142.0

We change back the 0 we assigned to values to np.NaN to avoid affecting any analysis we need to do.

In [82]:
change_back = rock['Release Year'] == 0
rock.loc[change_back, 'Release Year'] = np.NaN

In [83]:
rock['Release Year'].unique()

array([1982.,   nan, 1981., 1980., 1975., 2000., 2002., 1992., 1985.,
       1993., 1976., 1995., 1979., 1984., 1977., 1990., 1986., 1974.,
       2014., 1987., 1973., 2001., 1989., 1997., 1971., 1972., 1994.,
       1970., 1966., 1965., 1983., 1955., 1978., 1969., 1999., 1968.,
       1988., 1962., 2007., 1967., 1958., 1071., 1996., 1991., 2005.,
       2011., 2004., 2012., 2003., 1998., 2008., 1964., 2013., 2006.,
       1963., 1961.])

In [84]:
rock.head(20)

Song Clean   ARTIST CLEAN  Release Year  \
0              Caught Up in You    .38 Special        1982.0   
1                  Fantasy Girl    .38 Special           NaN   
2               Hold On Loosely    .38 Special        1981.0   
3        Rockin' Into the Night    .38 Special        1980.0   
4             Art For Arts Sake           10cc        1975.0   
5                    Kryptonite   3 Doors Down        2000.0   
6                         Loser   3 Doors Down        2000.0   
7                 When I'm Gone   3 Doors Down        2002.0   
8                    What's Up?  4 Non Blondes        1992.0   
9                    Take On Me           a-ha        1985.0   
10        Baby, Please Don't Go          AC/DC           NaN   
11                Back In Black          AC/DC        1980.0   
12                      Big Gun          AC/DC        1993.0   
13       CAN'T STOP ROCK'N'ROLL          AC/DC           NaN   
14  Dirty Deeds Done Dirt Cheap          AC/DC        1976.0   
15      For Those About To Rock          AC/DC        1981.0   
16             Girls Got Rhythm          AC/DC           NaN   
17               Hard As A Rock          AC/DC        1995.0   
18           Have a Drink On Me          AC/DC        1980.0   
19                  Hells Bells          AC/DC        1980.0   

                                 COMBINED  First?  Year?  PlayCount  F*G  
0         Caught Up in You by .38 Special       1      1         82   82  
1             Fantasy Girl by .38 Special       1      0          3    0  
2          Hold On Loosely by .38 Special       1      1         85   85  
3   Rockin' Into the Night by .38 Special       1      1         18   18  
4               Art For Arts Sake by 10cc       1      1          1    1  
5              Kryptonite by 3 Doors Down       1      1         13   13  
6                   Loser by 3 Doors Down       1      1          1    1  
7           When I'm Gone by 3 Doors Down       1      1          6    6  
8             What's Up? by 4 Non Blondes       1      1          3    3  
9                      Take On Me by a-ha       1      1          1    1  
10         Baby, Please Don't Go by AC/DC       1      0          1    0  
11                 Back In Black by AC/DC       1      1         97   97  
12                       Big Gun by AC/DC       1      1          6    6  
13        CAN'T STOP ROCK'N'ROLL by AC/DC       1      0          5    0  
14   Dirty Deeds Done Dirt Cheap by AC/DC       1      1         85   85  
15       For Those About To Rock by AC/DC       1      1         46   46  
16              Girls Got Rhythm by AC/DC       1      0         24    0  
17                Hard As A Rock by AC/DC       1      1          1    1  
18            Have a Drink On Me by AC/DC       1      1         39   39  
19                   Hells Bells by AC/DC       1      1         74   74

In [85]:
rock.describe(include=[np.number]).T

count         mean        std     min     25%     50%     75%  \
Release Year  1652.0  1978.019976  24.191247  1071.0  1971.0  1977.0  1984.0   
First?        2230.0     0.999552   0.021176     0.0     1.0     1.0     1.0   
Year?         2230.0     0.740807   0.438290     0.0     0.0     1.0     1.0   
PlayCount     2230.0    16.872197  25.303262     0.0     1.0     4.0    21.0   
F*G           2230.0    15.047982  25.288624     0.0     0.0     3.0    18.0   

                 max  
Release Year  2014.0  
First?           1.0  
Year?            1.0  
PlayCount      142.0  
F*G            142.0

In [86]:
rock.loc[rock['Release Year'] == 1071]

Song Clean ARTIST CLEAN  Release Year             COMBINED  First?  Year?  \
547      Levon   Elton John        1071.0  Levon by Elton John       1      1   

     PlayCount  F*G  
547          8    8

The minimum of the Release Year column is now 1071 and it is safe to assume this is an error. A simple google search shows this song was released in 1971 so we assume it's a data entry error where the 9 was replaced with a 0. We can change thsi back to 1971.

In [87]:
year_change = rock ['Release Year'] == 1071
rock.loc[year_change, 'Release Year'] = 1971

In [88]:
rock.describe(include=[np.number]).T

count         mean        std     min     25%     50%     75%  \
Release Year  1652.0  1978.564770   9.308823  1955.0  1971.0  1977.0  1984.0   
First?        2230.0     0.999552   0.021176     0.0     1.0     1.0     1.0   
Year?         2230.0     0.740807   0.438290     0.0     0.0     1.0     1.0   
PlayCount     2230.0    16.872197  25.303262     0.0     1.0     4.0    21.0   
F*G           2230.0    15.047982  25.288624     0.0     0.0     3.0    18.0   

                 max  
Release Year  2014.0  
First?           1.0  
Year?            1.0  
PlayCount      142.0  
F*G            142.0

In [94]:
rock.iloc[547]

Song Clean                    Levon
ARTIST CLEAN             Elton John
Release Year                 1971.0
COMBINED        Levon by Elton John
First?                            1
Year?                             1
PlayCount                         8
F*G                               8
Name: 547, dtype: object

In [179]:
rock.head(20)

Song Clean   ARTIST CLEAN  Release Year  \
0              Caught Up in You    .38 Special        1982.0   
1                  Fantasy Girl    .38 Special           NaN   
2               Hold On Loosely    .38 Special        1981.0   
3        Rockin' Into the Night    .38 Special        1980.0   
4             Art For Arts Sake           10cc        1975.0   
5                    Kryptonite   3 Doors Down        2000.0   
6                         Loser   3 Doors Down        2000.0   
7                 When I'm Gone   3 Doors Down        2002.0   
8                    What's Up?  4 Non Blondes        1992.0   
9                    Take On Me           a-ha        1985.0   
10        Baby, Please Don't Go          AC/DC           NaN   
11                Back In Black          AC/DC        1980.0   
12                      Big Gun          AC/DC        1993.0   
13       CAN'T STOP ROCK'N'ROLL          AC/DC           NaN   
14  Dirty Deeds Done Dirt Cheap          AC/DC        1976.0   
15      For Those About To Rock          AC/DC        1981.0   
16             Girls Got Rhythm          AC/DC           NaN   
17               Hard As A Rock          AC/DC        1995.0   
18           Have a Drink On Me          AC/DC        1980.0   
19                  Hells Bells          AC/DC        1980.0   

                                 COMBINED  First?  Year?  PlayCount  F*G  
0         Caught Up in You by .38 Special       1      1         82   82  
1             Fantasy Girl by .38 Special       1      0          3    0  
2          Hold On Loosely by .38 Special       1      1         85   85  
3   Rockin' Into the Night by .38 Special       1      1         18   18  
4               Art For Arts Sake by 10cc       1      1          1    1  
5              Kryptonite by 3 Doors Down       1      1         13   13  
6                   Loser by 3 Doors Down       1      1          1    1  
7           When I'm Gone by 3 Doors Down       1      1          6    6  
8             What's Up? by 4 Non Blondes       1      1          3    3  
9                      Take On Me by a-ha       1      1          1    1  
10         Baby, Please Don't Go by AC/DC       1      0          1    0  
11                 Back In Black by AC/DC       1      1         97   97  
12                       Big Gun by AC/DC       1      1          6    6  
13        CAN'T STOP ROCK'N'ROLL by AC/DC       1      0          5    0  
14   Dirty Deeds Done Dirt Cheap by AC/DC       1      1         85   85  
15       For Those About To Rock by AC/DC       1      1         46   46  
16              Girls Got Rhythm by AC/DC       1      0         24    0  
17                Hard As A Rock by AC/DC       1      1          1    1  
18            Have a Drink On Me by AC/DC       1      1         39   39  
19                   Hells Bells by AC/DC       1      1         74   74

## Write Some Functions

### Write a function that will take a row of a DataFrame and print out the song, artist, and whether or not the release date is < 1970

In [168]:
# ANSWER
def return_row(df, r):
    x = df.iloc[[r]]
    if x['Release Year'].item() < 1970:
        p = print('The song name is ' + str(x['Song Clean'].item()) + '. It was performed by ' + str(x['ARTIST CLEAN'].item()) + ' and it was released prior to 1970.')
    elif x['Release Year'].isna().item():
        p = print('The song name is ' + str(x['Song Clean'].item()) + '. It was performed by ' + str(x['ARTIST CLEAN'].item()) + ' and the release year is unknown.')
    else:
        p = print('The song name is ' + str(x['Song Clean'].item()) + '. It was performed by ' + str(x['ARTIST CLEAN'].item()) + ' and it was released after 1970.')
    return p

In [172]:
return_row(rock, 1)

The song name is Fantasy Girl. It was performed by .38 Special and the release year is unknown.


### Write a function that converts a column in a DataFrame to a numeric type and otherwise replaces entries with np.nan

In [174]:
# ANSWER
def convert_to_numeric(col):
    col = pd.to_numeric(col, errors='coerce')
    return col

### Apply this last function to your dataset

In [176]:
# ANSWER
rock[['Release Year', 'First?', 'Year?', 'PlayCount', 'F*G']].apply(convert_to_numeric)

Release Year  First?  Year?  PlayCount  F*G
0           1982.0       1      1         82   82
1              NaN       1      0          3    0
2           1981.0       1      1         85   85
3           1980.0       1      1         18   18
4           1975.0       1      1          1    1
...            ...     ...    ...        ...  ...
2225           NaN       1      0          1    0
2226        1981.0       1      1         32   32
2227        1975.0       1      1        109  109
2228        1983.0       1      1          1    1
2229        1973.0       1      1          2    2

[2230 rows x 5 columns]

### 'Describe' the new DataFrame

In [177]:
# ANSWER
rock.describe().T

count         mean        std     min     25%     50%     75%  \
Release Year  1652.0  1978.564770   9.308823  1955.0  1971.0  1977.0  1984.0   
First?        2230.0     0.999552   0.021176     0.0     1.0     1.0     1.0   
Year?         2230.0     0.740807   0.438290     0.0     0.0     1.0     1.0   
PlayCount     2230.0    16.872197  25.303262     0.0     1.0     4.0    21.0   
F*G           2230.0    15.047982  25.288624     0.0     0.0     3.0    18.0   

                 max  
Release Year  2014.0  
First?           1.0  
Year?            1.0  
PlayCount      142.0  
F*G            142.0

In [178]:
rock.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Song Clean    2230 non-null   object 
 1   ARTIST CLEAN  2230 non-null   object 
 2   Release Year  1652 non-null   float64
 3   COMBINED      2230 non-null   object 
 4   First?        2230 non-null   int64  
 5   Year?         2230 non-null   int64  
 6   PlayCount     2230 non-null   int64  
 7   F*G           2230 non-null   int64  
dtypes: float64(1), int64(4), object(3)
memory usage: 139.5+ KB


>




---



---



> > > > > > > > > © 2023 Institute of Data


---



---



